R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+#'
+#' @param object An object of the class age_model.
+setGeneric('get_parameters', function(object) standardGeneric('get_parameters'))
+
+setMethod(
+  'get_parameters', 'age_model',
+  function(object) object@parameters
+)
+
+#' @describeIn Sets parameters for an age-structured simple SEIR model.
+#'
+#' Default parameters a = 1, b = 1 and c = 1.
+#' If the initial conditions provided to do not sum to 1 or of different
+#' sizes compared to the number of age groups, an error is thrown.
+#'
+#' @param object An object of the class age_model.
+#' @param S0 initial fraction of the population that is susceptible
+#'           by age group.
+#' @param E0 initial fraction of the population that has been exposed
+#'           by age group.
+#' @param I0 initial fraction of the population that is infected
+#'           by age group.
+#' @param R0 initial fraction of the population that has recovered
+#'           by age group.
+#'
+#' All initial conditions must sum up to 1.
+#'
+#' @param a rate at which an infected individual exposes susceptible
+#' @param b rate at which exposed individuals become infected
+#' @param c rate at which infected individuals recover
+#'
+#' All rates of change between compartments are equal regardless of
+#' age group.
+#'
+#' @return updated version of the age-structured SEIR model.
+setGeneric(
+  'set_parameters',
+  function(object, S0, E0, I0, R0, a = 1, b = 1, c = 1){
+    standardGeneric('set_parameters')
+  })
+
+setMethod(
+  'set_parameters', 'age_model',
+  function(object, S0, E0, I0, R0, a, b, c) {
+
+    # check that ICs are valid
+    if (sum(S0, E0, I0, R0) != 1) {
+      stop('Invalid initial conditions. Must add up to 1.')
+    }
+
+    #create list of parameter values
+    params <- list(S0, E0, I0, R0, a, b, c)
+
+    #add names to each value
+    names(params) = object@parameter_names
+
+    #raise errors if age category dimensions do not match initial state vectors
+    #also raise errors if initial state and parameter values are not doubles
+    for (p in list('S0', 'E0', 'I0', 'R0')){
+      if(length(params[[p]]) != object@n_age_categories){
+        stop(glue('Wrong number of age groups for {p}
+              compartments.'))}
+      if(!is.double(params[[p]])){
+        stop(glue('{p} storage format must be a vector.'))}
+    }
+    if(sum(S0, E0, I0, R0) != 1){
+      stop('All comparmtnets need to sum up to 1.')
+    }
+
+    #check format of parameters a and b
+    if(any(length(a) != 1 | length(b) != 1 | length(c) != 1)){
+      stop('The rates of change between compartments are 1-dimensional.')
+    }
+
+    #if all above tests are passed, assign the params namelist to the object
+    object@parameters <- params
+
+    return(object)
+  })
+
+setGeneric(name = 'simulate',
+           def = function(object, times = seq(0, 100, by = 1), is_plot=TRUE){
+             standardGeneric('simulate')
+           }
+)
+
+#' @describeIn Solves an age-structured simple SEIR model.
```

Comment:
I'm not sure the "simple" is needed here.

Reply:
I think you may also want to include the method of solution used (i.e. numerical integration). I think it might also be good if we could allow a user to change the default numerical integration method used by deSolve (as think this might link quite well with some of the inference work that you're doing).

Reply:
Removed simple and described the ability of the class to solve the set of odes associated with the system through numerical integration of a chosen method. 

Also added argument to the simulate_SEIRAge function which is a string specifying the numerical solver. Default value is 'lsoda' which is the default for deSolve's ode function.

Reply:
Parfait, thanks.